Sources/SwiftSyntax/PlatformMutex.swift

```diff
@@ -0,0 +1,82 @@
+//===------ PlatformMutex.swift - Platform-specific Mutual Exclusion  -----===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#if canImport(Darwin)
+@_implementationOnly import Darwin
+#elseif canImport(Glibc)
+@_implementationOnly import Glibc
+#elseif canImport(WinSDK)
+@_implementationOnly import WinSDK
+#endif
+
+/// A protocol that platform-specific mutual exclusion primitives should conform to.
+struct PlatformMutex {
+  // FIXME: Use os_unfair_lock when we bump to macOS 12.0 on Darwin
```

Comment:
There's no pound-directive for macOS versions, but you can use `canImport` with a macOS 12+ module as a proxy, not sure how you feel about it.

```swift
#if canImport(GroupActivities) // GroupActivities was introduced in macOS 12.0
  typealias Primitive = os_unfair_lock
#elseif canImport(Darwin) || canImport(Glibc)
  typealias Primitive = pthread_mutex_t
#elseif canImport(WinSDK)
  typealias Primitive = SRWLOCK
#endif
```

Reply:
Awfully clever! But we're also looking into bumping our deployment targets anyways, so maybe we can roll onto macOS 12.0 et al here.

Reply:
I'll try out this branch with pthread_mutex and let you know if I can measure a meaningful difference.

Reply:
No bueno

```
swiftlint(14780,0x16dddf000) malloc: Heap corruption detected, free list is damaged at 0x60000281c3c0
*** Incorrect guard value: 0
swiftlint(14780,0x16dddf000) malloc: *** set a breakpoint in malloc_error_break to debug
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
```

Didn't have any issues with #807.

Reply:
I had a go with and without TSAN attached to swiftlint's test suite and I wasn't able to reproduce this. Do you have more detailed reproduction steps?

Reply:
This appears to be a revealed heap corruption that isn't in the parser or swift syntax. I got a bit suspicious since we don't ever `free` bump-allocated memory until the last red tree reference dies, which for swiftlint would only happen when the process is being torn down! 

```
Linting Swift files in current working directory
1 of 519 [                              ] ETA: 0s (11131 files/s)
The `anyobject_protocol` rule is now deprecated and will be completely removed in a future release.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
ThreadSanitizer:DEADLYSIGNAL
==41872==ERROR: ThreadSanitizer: SEGV on unknown address 0x0004051d8370 (pc 0x0001a9635c68 bp 0x00016df8d7d0 sp 0x00016df8d220 T7216288)
==41872==The signal is caused by a READ memory access.
    #0 cache_getImp <null>:130023616 (libobjc.A.dylib:arm64e+0x9c68) (BuildId: 4512d3f95f4d393fae7ffc2256226b4b32000000200000000100000000000d00)
    #1 _objc_msgSend_uncached <null>:130023616 (libobjc.A.dylib:arm64e+0x9b60) (BuildId: 4512d3f95f4d393fae7ffc2256226b4b32000000200000000100000000000d00)
    #2 CFStringGetBytes <null>:135270144 (CoreFoundation:arm64e+0x118e8) (BuildId: a075664168db330188211a2dd1efcf3732000000200000000100000000000d00)
    #3 CFStringAppend <null>:135270144 (CoreFoundation:arm64e+0x3399c) (BuildId: a075664168db330188211a2dd1efcf3732000000200000000100000000000d00)
    #4 -[NSPlaceholderMutableString initWithString:] <null>:135270064 (Foundation:arm64e+0xb0c4) (BuildId: 26a6775474d33ba882c1d341efdd3f3132000000200000000100000000000d00)
    #5 -[NSString stringByReplacingOccurrencesOfString:withString:options:range:] <null>:135270064 (Foundation:arm64e+0x2da14) (BuildId: 26a6775474d33ba882c1d341efdd3f3132000000200000000100000000000d00)
    #6 StringProtocol.replacingOccurrences<A, B>(of:with:options:range:) <null>:135270064 (Foundation:arm64e+0x426ea4) (BuildId: 26a6775474d33ba882c1d341efdd3f3132000000200000000100000000000d00)
    #7 FileNameRule.validate(file:) FileNameRule.swift:61 (swiftlint:arm64+0x1002f2948) (BuildId: 6285a9ceb9f23ad7bc3056e616504b2032000000200000000100000000000c00)
    #8 protocol witness for Rule.validate(file:using:compilerArguments:) in conformance FileNameRule <compiler-generated>:67 (swiftlint:arm64+0x10024ed68) (BuildId: 6285a9ceb9f23ad7bc3056e616504b2032000000200000000100000000000c00)
    #9 specialized Rule.lint(file:regions:benchmark:storage:configuration:superfluousDisableCommandRule:compilerArguments:) Linter.swift:84 (swiftlint:arm64+0x100204ef4) (BuildId: 6285a9ceb9f23ad7bc3056e616504b2032000000200000000100000000000c00)
    #10 closure #2 in CollectedLinter.getStyleViolations(using:benchmark:) Linter.swift:226 (swiftlint:arm64+0x1001fff2c) (BuildId: 6285a9ceb9f23ad7bc3056e616504b2032000000200000000100000000000c00)
    #11 partial apply for closure #2 in CollectedLinter.getStyleViolations(using:benchmark:) <compiler-generated> (swiftlint:arm64+0x100204a5c) (BuildId: 6285a9ceb9f23ad7bc3056e616504b2032000000200000000100000000000c00)
    #12 specialized closure #1 in closure #1 in Array.parallelMap<A>(transform:) Array+SwiftLint.swift:66 (swiftlint:arm64+0x1001d4bcc) (BuildId: 6285a9ceb9f23ad7bc3056e616504b2032000000200000000100000000000c00)
    #13 partial apply for specialized closure #1 in closure #1 in Array.parallelMap<A>(transform:) <compiler-generated> (swiftlint:arm64+0x1001d7034) (BuildId: 6285a9ceb9f23ad7bc3056e616504b2032000000200000000100000000000c00)
    #14 partial apply for thunk for @callee_guaranteed (@unowned Int) -> () <null>:129033504 (libswiftDispatch.dylib:arm64e+0x627c) (BuildId: c164cb0e54c03c9e9221235ef630f23332000000200000000100000000000d00)
    #15 thunk for @escaping @callee_guaranteed (@unowned Int) -> () <null>:129033504 (libswiftDispatch.dylib:arm64e+0x62a8) (BuildId: c164cb0e54c03c9e9221235ef630f23332000000200000000100000000000d00)
    #16 __wrap_dispatch_apply_block_invoke <null>:140518944 (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x8e2a4) (BuildId: 42d6e01b3578376b8cd36b02235d299832000000200000000100000000000b00)
    #17 _dispatch_client_callout2 <null>:129067008 (libdispatch.dylib:arm64e+0x4540) (BuildId: 2946ece1722a358abb26b063df295d7532000000200000000100000000000d00)
    #18 _dispatch_apply_invoke_and_wait <null>:129067008 (libdispatch.dylib:arm64e+0x19650) (BuildId: 2946ece1722a358abb26b063df295d7532000000200000000100000000000d00)
    #19 _dispatch_apply_with_attr_f <null>:129067008 (libdispatch.dylib:arm64e+0x18934) (BuildId: 2946ece1722a358abb26b063df295d7532000000200000000100000000000d00)
    #20 dispatch_apply <null>:129067008 (libdispatch.dylib:arm64e+0x18b44) (BuildId: 2946ece1722a358abb26b063df295d7532000000200000000100000000000d00)
    #21 wrap_dispatch_apply <null>:140518944 (libclang_rt.tsan_osx_dynamic.dylib:arm64e+0x8e214) (BuildId: 42d6e01b3578376b8cd36b02235d299832000000200000000100000000000b00)
    #22 _swift_dispatch_apply_current <null>:129033504 (libswiftDispatch.dylib:arm64e+0x634c) (BuildId: c164cb0e54c03c9e9221235ef630f23332000000200000000100000000000d00)
    #23 static OS_dispatch_queue.concurrentPerform(iterations:execute:) <null>:129033504 (libswiftDispatch.dylib:arm64e+0x6200) (BuildId: c164cb0e54c03c9e9221235ef630f23332000000200000000100000000000d00)
    #24 specialized closure #1 in Array.parallelMap<A>(transform:) Array+SwiftLint.swift:65 (swiftlint:arm64+0x1001d47b0) (BuildId: 6285a9ceb9f23ad7bc3056e616504b2032000000200000000100000000000c00)
    #25 specialized Array.parallelCompactMap<A>(transform:) Array+SwiftLint.swift:59 (swiftlint:arm64+0x1001d3f1c) (BuildId: 6285a9ceb9f23ad7bc3056e616504b2032000000200000000100000000000c00)
    #26 CollectedLinter.getStyleViolations(using:benchmark:) Linter.swift:225 (swiftlint:arm64+0x1001fef00) (BuildId: 6285a9ceb9f23ad7bc3056e616504b2032000000200000000100000000000c00)
    #27 CollectedLinter.styleViolations(using:) Linter.swift:197 (swiftlint:arm64+0x1001fec64) (BuildId: 6285a9ceb9f23ad7bc3056e616504b2032000000200000000100000000000c00)
    #28 closure #1 in static LintOrAnalyzeCommand.collectViolations(builder:) LintOrAnalyzeCommand.swift:70 (swiftlint:arm64+0x100faaba8) (BuildId: 6285a9ceb9f23ad7bc3056e616504b2032000000200000000100000000000c00)
    #29 swift::runJobInEstablishedExecutorContext(swift::Job*) <null>:129030240 (libswift_Concurrency.dylib:arm64e+0x41480) (BuildId: b0a6b4dfc78c3118a7bd362d673eb5dd32000000200000000100000000000b00)
```

Reply:
Oh wait, never mind, that stack is absolutely bogus.